### ISSUE TITLE
Graphical issues on Yoshi's Island

### ISSUE THREAD

**Issue by TrashRaccoonSnuffy** [xyproto/zsnes#18]
![image](https://github.com/xyproto/zsnes/assets/127693805/dfdbfdb1-78ba-4542-a900-dd345dd57a79)
![image](https://github.com/xyproto/zsnes/assets/127693805/e0117d26-4b4d-4ed9-8037-74f96d5b36e9)
![image](https://github.com/xyproto/zsnes/assets/127693805/68b92e94-78f8-4e30-adc4-164b4b6f1644)

Most likely related to improper HDMA/DMA emulation (or the code being converted wrongly for 64-bit?), the game looks fine on ZSNES 1.51 and 1.42

**Comment by xyproto** [comment]
Thanks for reporting!

**Comment by TrashRaccoonSnuffy** [comment]
I've been investigating this issue myself and this seems to be a problem in c_reg420Cw in cpu/c_dma.c, Specifically the
`
nexthdma & al == 0
`
part in the if statement.

If I remove that, the HDMA works as expected and the graphics bugs are gone. Yoshi's Island does 2 writes to 420C (The HDMA Enable register), one with 00 and one with the actual expected value. A per-game hack will probably have to be done to fix this, since this issue was added after 1.51 when regs got ported to C. There might have been a wrong conversion somewhere in the code. I'll make a fix for it.

The other secondary fix I found was setting per2exec in zsnesl.cfg to 50, but this makes the game run very slowly and flicker.
